docs/proposals/user-namespace.md

```diff
@@ -0,0 +1,206 @@
+leaves the container runtimes free to manage the permissions of the 
+directories under the pod directory (available via injecting the
+`RuntimeHelper` into the runtime implementation) in a way that suits
+their ability to provide the GID.  For Docker this is the value in
+`/etc/subgid`.
+
+There is movement towards being able to [specify](https://github.com/docker/docker/issues/28593) 
+or [lock ranges](https://github.com/coreos/rkt/issues/1090)
+of UIDs/GIDs in the runtime community.  As this work progresses it 
+makes sense that the ability to allocate a range should belong to the
+orchestration system.  This would allow Kuberenetes to know exactly
+what GID requires access to the directories and ensure the correct
+access is granted.  
+
+This proposal recommends that, as a next step in support, the system
+could assume that the `FSGroup` represents the correct GID.  When
+the node is in user namespace remap mode (known by the existing flag)
+the runtime can use the `FSGroup` to grant access to the 
+`/var/lib/kubelet/pods/{UID}/...` subdirectories.
+
+The `FSGroup` will still function normally for existing volume types
+that support it by changing the group ownership.  This will allow
+usage of those volume types in the remap environment.
+
+
+### Escaping Private User Namespaces
+
+Some containers may use features which prohibit the use of private
+user namespaces.  If the Kubelet is aware that private user namespaces
+are in use it has the ability to modify the container's runtime 
+configuration when it detects a feature that is not compatible with
+private user namespaces.
+
+If the system is to default host user namespace options it would occur
+when:
+
+1.  setting any other namespace flag to host.  Not strictly
+required for net=host if not binding to low ports but if the system
+provides defaulting it covers the 80% case
+1.  using capabilities that aren't tied to any namespace
+    * MKNOD
+    * SYS_MODULE
+    * SYS_TIME
+1.  explicitly granted permissions to use a host volume and escape the private user namespace
+
+This logic was implemented in the [experimental
+support](https://github.com/kubernetes/kubernetes/pull/31169) as of
+Kubernetes 1.5.  
+
+Open Question: should we allow the usage of host user namespace as a 
+flag in addition to defaulting?  Ie. exposing `HostUserNamespace` as 
+a flag on the `SecurityContext` like we do with net, ipc, and pid?
+
+### Enforcing Private User Namespaces
+
+In some cases a cluster administrator may wish to ensure that a container
+is never defaulted to use the host user namespace.  This falls under
+the responsibility of cluster policy with `PodSecurityPolicy`.  
+
+`PodSecurityPolicy` has the ability to control using the host network,
+PID, and IPC namespaces, which capabilities can be added, as well as
+which types of volumes can be used.  This effectively controls who
+will be allowed to trigger the Kubelet defaulting logic that would 
+escape private user namespaces.  
```

Comment:
@pweil- I am not sure even if using PSP will help me in the case of hostpath. As mentioned earlier, in distributed file system case where DFS is mounted on each underlying host. I need to allow users to be able to use PVC which is backed by hostpath. The current design will then escape `usernamespace`. I would really like to have an option in PSP or in pod spec security context which enables me choose to honor defaulting logic or override it. Now as the defaulting logic is already implemented in 1.5 and we are talking about phase2, I think this is quite necessary to have such an option for users like me who has DFS and want to use usernamespaces.